File: scalyc/src/compiler/parser/lexer.rs
use std::collections::HashSet;
use std::str::Chars;

pub struct Position {
    pub line: usize,
    pub column: usize,
}

pub enum Token {
    Empty,
    Invalid,
    Identifier(String),
    Attribute(String),
    Literal(Literal),
    Punctuation(String),
    LineFeed,
    Colon,
    Semicolon,
}

pub enum Literal {
    StringLiteral(String),
    Fragment(String),
    Integer(String),
    FloatingPoint(String),
    Hex(String),
}

pub struct Lexer<'a> {
    pub token: Token,
    character: Option<char>,
    chars: Chars<'a>,
    previous_line: usize,
    previous_column: usize,
    pub line: usize,
    pub column: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(deck: &'a str) -> Lexer {
        let mut lexer = Lexer {
            chars: deck.chars(),
            token: Token::Empty,
            character: None,
            previous_line: 1,
            previous_column: 0,
            line: 1,
            column: 0,
        };
        lexer.read_character();
        lexer
    }

    fn read_character(&mut self) {
        self.character = self.chars.next();
        match self.character {
            None => (),
            Some(c) => match c {
                '\r' => (),
                '\n' => {
                    self.column = 0;
                    self.line = self.line + 1;
                }
                '\t' => self.column = self.column + 4,
                _ => self.column = self.column + 1,
            },
        }
    }

    pub fn advance(&mut self) {
        self.skip_whitespace(false);
        self.previous_line = self.line;
        self.previous_column = self.column;
        match self.character {
            None => return,
            Some(c) => {
                if ((c >= 'a') && (c <= 'z')) || ((c >= 'A') && (c <= 'Z')) {
                    self.token = self.scan_identifier();
                    return;
                }

                if (c >= '1') && (c <= '9') {
                    let mut value = String::new();
                    value.push(c);
                    self.token = self.scan_numeric_literal(value);
                    return;
                }

                match c {
                    '\n' => {
                        self.token = self.scan_line_feed();
                    }
                    ':' => {
                        self.read_character();
                        self.token = Token::Colon;
                    }
                    ';' => {
                        self.read_character();
                        self.token = Token::Semicolon;
                    }
                    '0' => {
                        let mut value = String::new();
                        value.push(c);
                        self.token = self.scan_numeric_or_hex_literal(value);
                    }
                    '@' => {
                        self.read_character();
                        self.token = self.scan_attribute();
                        return;
                    }
                    '+' | '-' | '*' | '/' | '=' | '%' | '&' | '|' | '^' | '~' | '<' | '>' => {
                        let mut operation = String::new();
                        operation.push(c);
                        self.token = self.scan_operator(operation)
                    }
                    '\"' => self.token = { self.scan_string_literal() },
                    '\'' => self.token = { self.scan_string_identifier() },
                    '`' => self.token = { self.scan_fragment_literal() },
                    '}' | ')' | ']' | '.' | '?' => {
                        let mut punctuation_string = String::new();
                        punctuation_string.push(c);
                        self.read_character();
                        self.token = Token::Punctuation(punctuation_string);
                    }
                    '{' | '(' | '[' | ',' => {
                        let mut punctuation_string = String::new();
                        punctuation_string.push(c);
                        self.read_character();
                        self.skip_whitespace(true);
                        self.token = Token::Punctuation(punctuation_string);
                    }
                    _ => self.token = Token::Invalid,
                }
            }
        }
    }

    pub fn empty(&mut self) {
        self.token = Token::Empty;
    }

    fn scan_line_feed(&mut self) -> Token {
        loop {
            self.read_character();
            self.skip_whitespace(false);
            match self.character {
                None => {
                    return Token::LineFeed;
                }

                Some(c) => match c {
                    '\n' => continue,
                    _ => return Token::LineFeed,
                },
            }
        }
    }

    fn scan_identifier(&mut self) -> Token {
        let mut name = String::new();
        loop {
            match self.character {
                None => {
                    if name.len() == 0 {
                        return Token::Invalid;
                    } else {
                        return Token::Identifier(name);
                    }
                }

                Some(c) => match c {
                    'a'..='z' | 'A'..='Z' | '0'..='9' | '_' => {
                        name.push(c);
                        self.read_character();
                    }
                    _ => return Token::Identifier(name),
                },
            }
        }
    }

    fn scan_attribute(&mut self) -> Token {
        let mut name = String::new();
        loop {
            match self.character {
                None => {
                    if name.len() == 0 {
                        return Token::Invalid;
                    } else {
                        return Token::Attribute(name);
                    }
                }

                Some(c) => match c {
                    'a'..='z' | 'A'..='Z' | '0'..='9' | '_' => {
                        name.push(c);
                        self.read_character();
                    }
                    _ => return Token::Attribute(name),
                },
            }
        }
    }

    fn scan_operator(&mut self, mut operation: String) -> Token {
        loop {
            match self.character {
                None => {
                    if operation.len() == 0 {
                        return Token::Invalid;
                    } else {
                        return Token::Identifier(operation);
                    }
                }
                Some(c) => match c {
                    '+' | '-' | '*' | '/' | '=' | '%' | '&' | '|' | '^' | '~' | '<' | '>' => {
                        operation.push(c);
                        self.read_character();
                    }

                    _ => return Token::Identifier(operation),
                },
            }
        }
    }

    fn scan_string_literal(&mut self) -> Token {
        let mut value = String::new();

        loop {
            self.read_character();

            match self.character {
                None => return Token::Invalid,
                Some(c) => match c {
                    '\"' => {
                        self.read_character();
                        return Token::Literal(Literal::StringLiteral(value));
                    }

                    '\\' => {
                        self.read_character();
                        match self.character {
                            None => return Token::Invalid,
                            Some(c) => match c {
                                '\"' | '\\' | '\'' => {
                                    value.push('\\');
                                    value.push(c);
                                }
                                'n' => {
                                    value.push('\\');
                                    value.push('n');
                                }
                                'r' => {
                                    value.push('\\');
                                    value.push('r');
                                }
                                't' => {
                                    value.push('\\');
                                    value.push('t');
                                }
                                '0' => {
                                    value.push('\\');
                                    value.push('0');
                                }
                                _ => return Token::Invalid,
                            },
                        }
                    }

                    _ => {
                        value.push(c);
                    }
                },
            }
        }
    }

    fn scan_string_identifier(&mut self) -> Token {
        let mut value = String::new();

        loop {
            self.read_character();

            match self.character {
                None => return Token::Invalid,
                Some(c) => match c {
                    '\'' => {
                        self.read_character();
                        return Token::Identifier(value);
                    }

                    _ => {
                        value.push(c);
                    }
                },
            }
        }
    }

    fn scan_fragment_literal(&mut self) -> Token {
        let mut value = String::new();

        loop {
            self.read_character();

            match self.character {
                None => return Token::Invalid,
                Some(c) => match c {
                    '`' => {
                        self.read_character();
                        return Token::Literal(Literal::Fragment(value));
                    }

                    '\\' => {
                        self.read_character();
                        match self.character {
                            None => return Token::Invalid,
                            Some(c) => match c {
                                '`' => {
                                    value.push(c);
                                }
                                _ => {
                                    value.push('\\');
                                    value.push(c);
                                }
                            },
                        }
                    }

                    _ => {
                        value.push(c);
                    }
                },
            }
        }
    }

    fn scan_numeric_or_hex_literal(&mut self, mut value: String) -> Token {
        self.read_character();
        match self.character {
            None => return Token::Literal(Literal::Integer(value)),
            Some(c) => match c {
                '0'..='9' => {
                    value.push(c);
                    return self.scan_numeric_literal(value);
                }
                '.' => {
                    value.push(c);
                    return self.scan_fraction(value);
                }
                'E' | 'e' => {
                    value.push(c);
                    return self.scan_exponent(value);
                }
                'x' => {
                    return self.scan_hex_literal();
                }
                _ => return Token::Literal(Literal::Integer(value)),
            },
        }
    }

    fn scan_numeric_literal(&mut self, mut value: String) -> Token {
        loop {
            self.read_character();
            match self.character {
                None => {
                    return Token::Literal(Literal::Integer(value));
                }
                Some(c) => match c {
                    '0'..='9' => {
                        value.push(c);
                    }
                    '.' => {
                        value.push(c);
                        return self.scan_fraction(value);
                    }
                    'E' | 'e' => {
                        value.push(c);
                        return self.scan_exponent(value);
                    }
                    _ => return Token::Literal(Literal::Integer(value)),
                },
            }
        }
    }

    fn scan_fraction(&mut self, mut value: String) -> Token {
        loop {
            self.read_character();
            match self.character {
                None => {
                    return Token::Literal(Literal::FloatingPoint(value));
                }
                Some(c) => match c {
                    '0'..='9' => {
                        value.push(c);
                    }
                    'E' | 'e' => {
                        value.push(c);
                        return self.scan_exponent(value);
                    }
                    _ => return Token::Literal(Literal::FloatingPoint(value)),
                },
            }
        }
    }

    fn scan_exponent(&mut self, mut value: String) -> Token {
        loop {
            self.read_character();
            match self.character {
                None => {
                    return Token::Literal(Literal::FloatingPoint(value));
                }
                Some(c) => match c {
                    '0'..='9' => {
                        value.push(c);
                    }
                    _ => return Token::Literal(Literal::FloatingPoint(value)),
                },
            }
        }
    }

    fn scan_hex_literal(&mut self) -> Token {
        let mut value = String::new();
        loop {
            self.read_character();
            match self.character {
                None => {
                    return Token::Literal(Literal::FloatingPoint(value));
                }
                Some(c) => match c {
                    '0'..='9' | 'a'..='f' | 'A'..='F' => {
                        value.push(c);
                    }
                    _ => return Token::Literal(Literal::Hex(value)),
                },
            }
        }
    }

    fn skip_whitespace(&mut self, line_feed: bool) {
        loop {
            match self.character {
                None => return,
                Some(c) => match c {
                    ' ' => {
                        self.read_character();
                        continue
                    }
                    '\t' => {
                        self.read_character();
                        continue
                    }
                    '\r' => {
                        self.read_character();
                        continue
                    }
                    '\n' => if line_feed {
                        self.read_character();
                        continue
                    }
                    else {
                        return
                    }
                    '/' => {
                        self.read_character();
                        match self.character {
                            None => return,
                            Some(c) => match c {
                                '/' => {
                                    self.read_character();
                                    self.handle_single_line_comment()
                                }
                                '*' => {
                                    self.read_character();
                                    self.handle_multi_line_comment()
                                }
                                _ => return,
                            },
                        }
                    }

                    _ => {
                        return
                    }
                },
            }
        }
    }

    fn handle_single_line_comment(&mut self) {
        loop {
            match self.character {
                None => return,
                Some(c) => match c {
                    '\t' => {
                        self.read_character();
                        continue;
                    }
                    '\r' => {
                        continue;
                    }
                    '\n' => {
                        self.read_character();
                        return;
                    }
                    _ => {
                        self.read_character();
                        continue;
                    }
                },
            }
        }
    }

    fn handle_multi_line_comment(&mut self) {
        loop {
            match self.character {
                None => return,
                Some(c) => match c {
                    '/' => {
                        self.read_character();
                        match self.character {
                            None => return,
                            Some(c) => match c {
                                '*' => self.handle_multi_line_comment(),
                                _ => return,
                            },
                        }
                    }

                    '*' => {
                        self.read_character();
                        match self.character {
                            None => return,
                            Some(c) => match c {
                                '/' => {
                                    self.read_character();
                                    return;
                                }
                                _ => (),
                            },
                        }
                    }

                    '\t' => {
                        self.read_character();
                        continue;
                    }

                    '\r' => {
                        continue;
                    }

                    '\n' => {
                        self.read_character();
                        continue;
                    }

                    _ => {
                        self.read_character();
                        continue;
                    }
                },
            }
        }
    }

    pub fn parse_keyword(&mut self, fixed_string: String) -> bool {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Identifier(name) => {
                let right_keyword = *name == fixed_string;
                if right_keyword {
                    self.empty();
                }
                return right_keyword;
            }
            _ => return false,
        }
    }

    pub fn parse_identifier(&mut self, keywords: &HashSet<String>) -> Option<String> {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Identifier(name) => {
                if keywords.contains(name) {
                    return None;
                }
                let ret = String::from(name);
                self.empty();
                return Some(ret);
            }
            _ => return None,
        }
    }

    pub fn parse_attribute(&mut self) -> Option<String> {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Attribute(name) => {
                let ret = String::from(name);
                self.empty();
                return Some(ret);
            }
            _ => return None,
        }
    }

    pub fn parse_punctuation(&mut self, fixed_string: String) -> bool {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Punctuation(name) => {
                let ret = *name == fixed_string;
                if ret {
                    self.empty();
                }
                return ret;
            }
            _ => return false,
        }
    }

    pub fn parse_literal(&mut self) -> Option<Literal> {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Literal(Literal::StringLiteral(name)) => {
                let ret = String::from(name);
                self.empty();
                Some(Literal::StringLiteral(ret))
            }
            Token::Literal(Literal::Integer(name)) => {
                let ret = String::from(name);
                self.empty();
                Some(Literal::Integer(ret))
            }
            Token::Literal(Literal::FloatingPoint(name)) => {
                let ret = String::from(name);
                self.empty();
                Some(Literal::FloatingPoint(ret))
            }
            Token::Literal(Literal::Hex(name)) => {
                let ret = String::from(name);
                self.empty();
                Some(Literal::Hex(ret))
            }
            _ => None,
        }
    }

    pub fn parse_colon(&mut self) -> bool {
        match &self.token {
            Token::Empty => self.advance(),
            _ => (),
        }
        match &self.token {
            Token::Colon | Token::LineFeed => {
                self.empty();
                true
            }
            _ => false,
        }
    }

    pub fn is_at_end(&mut self) -> bool {
        match self.character {
            None => true,
            Some(_) => false,
        }
    }

    pub fn get_position(&self) -> Position {
        Position {
            line: self.line,
            column: self.column,
        }
    }

    pub fn get_previous_position(&self) -> Position {
        Position {
            line: self.previous_line,
            column: self.previous_column,
        }
    }
}

File: Scaly.Compiler/Lexer.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Scaly.Compiler
{
    class Lexer
    {
        public Token token;
        char? character;
        CharEnumerator chars;
        public ulong previous_line;
        public ulong previous_column;
        public ulong line;
        public ulong column;

        public Lexer(string deck)
        {
            token = new EmptyToken();
            chars = deck.GetEnumerator();
            character = null;
            previous_line = 1;
            previous_column = 0;
            line = 1;
            column = 0;
            read_character();
            skip_whitespace(true);
        }

        void read_character()
        {
            if (chars.MoveNext())
                character = chars.Current;
            else
                character = null;

            switch (character)
            {
                case '\r':
                    break;

                case '\n':
                    column = 0;
                    line += 1;
                    break;

                default:
                    column++;
                    break;
            }
        }

        public void advance()
        {
            skip_whitespace(false);
            previous_line = line;
            previous_column = column;

            if (character == null)
                return;

            char c = (char)character;

            if (((c >= 'a') && (c <= 'z')) || ((c >= 'A') && (c <= 'Z')))
            {
                token = scan_identifier();
                return;
            }

            if ((c >= '1') && (c <= '9'))
            {
                var value = new StringBuilder();
                value.Append(c);
                token = scan_numeric_literal(value);
                return;
            }

            switch (c)
            {
                case '\n':
                    token = scan_line_feed();
                    break;

                case ':':
                    read_character();
                    token = new Colon();
                    break;

                case ';':
                    read_character();
                    token = new Semicolon();
                    break;

                case '0':
                    {
                        var value = new StringBuilder();
                        value.Append(c);
                        token = scan_numeric_or_hex_literal(value);
                    }
                    break;

                case '@':
                    read_character();
                    token = scan_attribute();
                    break;

                case '+': case '-': case '*': case '/': case '=': case '%': case '&': case '|': case '^': case '~': case '<': case '>':
                {
                    var operation = new StringBuilder();
                    operation.Append(c);
                    token = scan_operator(operation);
                    break;
                }

                case '\"': token = scan_string_literal(); break;
                case '\'': token = scan_string_identifier(); break;
                case '`':  token = scan_fragment_literal(); break;

                case '}': case ')': case ']': case '.': case '?':
                {
                    var punctuation_string = new StringBuilder();
                    punctuation_string.Append(c);
                    read_character();
                    token = new Punctuation(punctuation_string.ToString());
                    break;
                }

                case '{': case '(': case '[': case ',':
                {
                    var punctuation_string = new StringBuilder();
                    punctuation_string.Append(c);
                    read_character();
                    skip_whitespace(true);
                    token = new Punctuation(punctuation_string.ToString());
                    break;
                }

                default:
                    token = new InvalidToken();
                    break;
            }
        }

        public void empty() {
            token = new EmptyToken();
        }

        Token scan_line_feed()
        {
            while (true)
            {
                read_character();
                skip_whitespace(false);
                if (character == null)
                    return new InvalidToken();

                if (character == '\n')
                    continue;

                return new LineFeed();
            }
        }

        Token scan_identifier()
        {
            var name = new StringBuilder();
            while (true)
            {
                if (character == null)
                {
                    if (name.Length == 0)
                        return new InvalidToken();
                    else
                        return new Identifier(name.ToString());
                }
                else
                {
                    char c = (char)character;
                    if (((c >= 'a') && (c <= 'z')) || ((c >= 'A') && (c <= 'Z')) || ((c >= '0') && (c <= '9')) || (c == '_'))
                    {
                        name.Append(c);
                        read_character();
                    }
                    else
                    {
                        return new Identifier(name.ToString());
                    }
                }
            }
        }

        Token scan_attribute()
        {
            var name = new StringBuilder();
            while (true)
            {
                if (character == null)
                {
                    if (name.Length == 0)
                        return new InvalidToken();
                    else
                        return new Attribute(name.ToString());
                }

                var c = (char)character;
                if (((c >= 'a') && (c <= 'z')) || ((c >= 'A') && (c <= 'Z')) || ((c >= '0') && (c <= '9')) || (c == '_'))
                {
                    name.Append(c);
                    read_character();
                }
                else
                {
                    return new Attribute(name.ToString());
                }
            }
        }

        Token scan_operator(StringBuilder operation)
        {
            while (true)
            {
                read_character();
                if (character == null)
                {
                    if (operation.Length == 0)
                        return new InvalidToken();
                    else
                        return new Identifier(operation.ToString());
                }

                var c = (char)character;
                switch (c)
                {
                    case '+': case '-': case '*': case '/': case '=': case '%': case '&': case '|': case '^': case '~': case '<': case '>':
                        operation.Append(c);
                        break;

                    default:
                        return new Identifier(operation.ToString());
                }
            }
        }

        Token scan_string_literal()
        {
            var value = new StringBuilder();

            while (true)
            {
                read_character();

                if (character == null)
                    return new InvalidToken();

                var c = (char)character;
                switch (c)
                {
                    case '\"':
                        read_character();
                        return new StringLiteral(value.ToString());

                    case '\\':
                        read_character();
                        if (character == null)
                            return new InvalidToken();

                        var c2 = (char)character;
                        switch (c2)
                        {
                            case '\"': case '\\': case '\'':
                                value.Append('\\');
                                value.Append(c2);
                                break;
                            case 'n':
                                value.Append('\\');
                                value.Append('n');
                                break;
                            case 'r':
                                value.Append('\\');
                                value.Append('r');
                                break;
                            case 't':
                                value.Append('\\');
                                value.Append('t');
                                break;
                            case '0':
                                value.Append('\\');
                                value.Append('0');
                                break;
                            default:
                                return new InvalidToken();
                        }
                        break;

                    default:
                        value.Append(c);
                        break;
                }
            }
        }

        Token scan_string_identifier()
        {
            var value = new StringBuilder();

            while (true)
            {
                read_character();

                if (character == null)
                    return new InvalidToken();

                var c = (char)character;
                switch (c)
                {
                    case '\'':
                        read_character();
                        return new Identifier(value.ToString());
                    default:
                        value.Append(c);
                        break;
                }
            }
        }

        Token scan_fragment_literal()
        {
            var value = new StringBuilder();

            while (true)
            {
                read_character();

                if (character == null)
                    return new InvalidToken();

                var c = (char)character;
                switch (c)
                {
                    case '`':
                        read_character();
                        return new Fragment(value.ToString());

                    case '\\':
                        read_character();
                        if (character == null)
                            return new InvalidToken();

                        var c2 = (char)character;
                        switch (c2)
                        {
                            case '`':
                                value.Append(c2);
                                break;
                            default:
                                value.Append('\\');
                                value.Append(c2);
                                break;
                        }
                        break;

                    default:
                        value.Append(c);
                        break;
                }
            }
        }

        Token scan_numeric_or_hex_literal(StringBuilder value)
        {
            read_character();

            if (character == null)
                return new InvalidToken();

            var c = (char)character;

            if ((c >= '0') && (c <= '9'))
            {
                value.Append(c);
                return scan_numeric_literal(value);
            }

            switch (c)
            {
                case '.':
                    value.Append(c);
                    return scan_fraction(value);

                case 'E':
                case 'e':
                    value.Append(c);
                    return scan_exponent(value);

                case 'x':
                    return scan_hex_literal();

                default:
                    return new Integer(value.ToString());
            }
        }

        Token scan_numeric_literal(StringBuilder value)
        {
            while (true)
            {
                read_character();

                if (character == null)
                    return new Integer(value.ToString());

                var c = (char)character;

                if ((c >= '0') && (c <= '9'))
                {
                    value.Append(c);
                    continue;
                }

                switch (c)
                {
                    case '.':
                        value.Append(c);
                        return scan_fraction(value);

                    case 'E':
                    case 'e':
                        value.Append(c);
                        return scan_exponent(value);

                    default:
                        return new Integer(value.ToString());
                }
            }
        }

        Token scan_fraction(StringBuilder value)
        {
            while (true)
            {
                read_character();

                if (character == null)
                    return new FloatingPoint(value.ToString());

                var c = (char)character;

                if ((c >= '0') && (c <= '9'))
                {
                    value.Append(c);
                    continue;
                }

                switch (c)
                {
                    case 'E':
                    case 'e':
                        value.Append(c);
                        return scan_exponent(value);

                    default:
                        return new FloatingPoint(value.ToString());
                }
            }
        }

        Token scan_exponent(StringBuilder value)
        {
            while (true)
            {
                read_character();

                if (character == null)
                    return new FloatingPoint(value.ToString());

                var c = (char)character;

                if ((c >= '0') && (c <= '9'))
                {
                    value.Append(c);
                    continue;
                }

                return new FloatingPoint(value.ToString());
            }
        }

        Token scan_hex_literal()
        {
            var value = new StringBuilder();
            while (true)
            {
                read_character();

                if (character == null)
                    return new Hex(value.ToString());

                var c = (char)character;

                if (((c >= '0') && (c <= '9')) || (c >= 'A') && (c <= 'F') || ((c >= 'a') && (c <= 'f')))
                {
                    value.Append(c);
                    continue;
                }

                return new Hex(value.ToString());
            }
        }

        void skip_whitespace(bool line_feed)
        {
            while (true)
            {
                if (character == null)
                    return;

                var c = (char)character;

                switch(c)
                {
                    case ' ':
                        read_character();
                        continue;
                    case '\t':
                        read_character();
                        continue;
                    case '\r':
                        read_character();
                        continue;
                    case '\n':
                        if (line_feed)
                        {
                            read_character();
                            continue;
                        }
                        else
                        {
                            return;
                        }

                    case ';':
                        read_character();
                        if (character == null)
                            return;

                        var c2 = (char)character;
                        switch (c2)
                        {
                            case '*':
                                read_character();
                                handle_multi_line_comment();
                                break;

                            default:
                                read_character();
                                handle_single_line_comment();
                                break;
                        }
                        break;
                    default:
                        return;
                }
            }
        }


        void handle_single_line_comment()
        {
            while (true)
            {
                if (character == null)
                    return;

                var c = (char)character;

                switch (c)
                {
                    case '\t':
                        read_character();
                        continue;

                    case '\r':
                        read_character();
                        continue;

                    case '\n':
                        read_character();
                        return;

                    default:
                        read_character();
                        continue;
                }
            }
        }

        void handle_multi_line_comment()
        {
            while (true)
            {
                if (character == null)
                    return;
                var c = (char)character;

                switch (c)
                {
                    case ';':
                        read_character();
                        if (character == null)
                            return;
                        if (character == '*')
                            handle_multi_line_comment();
                        else
                            continue;
                        break;

                    case '*':
                        read_character();
                        if (character == null)
                            return;

                        c = (char)character;
                        if (c == ';')
                        {
                            read_character();
                            return;
                        }
                        break;

                    default:
                        read_character();
                        continue;
                }
            }
        }

        public bool parse_keyword(string fixed_string)
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case Identifier identifier:
                    var right_keyword = (identifier.name == fixed_string);
                    if (right_keyword)
                        empty();

                    return right_keyword;

                default:
                    return false;
            }
        }

        public string parse_identifier(HashSet<string> keywords)
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case Identifier identifier:
                    if (keywords.Contains(identifier.name))
                        return null;
                    empty();
                    return identifier.name;

                default:
                    return null;
            }
        }

        public string parse_attribute()
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case Attribute attribute:
                    empty();
                    return attribute.name;

                default:
                    return null;
            }
        }

        public bool parse_punctuation(string fixed_string)
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case Punctuation punctuation:
                    var ret = (punctuation.sign == fixed_string);
                    if (ret)
                        empty();

                    return ret;
                default:
                    return false;
            }
        }

        public Literal parse_literal()
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case StringLiteral stringLiteral:
                    empty();
                    return stringLiteral;

                case Integer integer:
                    empty();
                    return integer;

                case FloatingPoint floatingPoint:
                    empty();
                    return floatingPoint;

                case Hex hex:
                    empty();
                    return hex;

                case Fragment fragment:
                    empty();
                    return fragment;

                default:
                    return null;
            }
        }

        public bool parse_colon()
        {
            if (token is EmptyToken)
                advance();

            switch (token)
            {
                case Colon: case LineFeed:
                    empty();
                    return true;

                default:
                    return false;
            }
        }

        public bool is_at_end()
        {
            if (character == null)
                return true;

            return false;
        }

        public Position get_position()
        {
            return new Position
            {
                line = line,
                column = column,
            };
        }

        public Position get_previous_position()
        {
            return new Position
            {
                line = previous_line,
                column = previous_column,
            };
        }
    }

    public class Token
    {
    }

    public class EmptyToken : Token
    {
    }

    public class InvalidToken : Token
    {
    }

    public class Identifier : Token
    {
        public string name;
        public Identifier(string name)
        {
            this.name = name;
        }
    }

    public class Attribute : Token
    {
        public string name;
        public Attribute(string name)
        {
            this.name = name;
        }
    }

    public class Punctuation : Token
    {
        public string sign;
        public Punctuation(string theSign)
        {
            sign = theSign;
        }
    }

    public class Literal : Token
    {
    }

    public class StringLiteral : Literal
    {
        public string value;
        public StringLiteral(string theString)
        {
            value = theString;
        }
    }

    public class Fragment : Literal
    {
        public string value;
        public Fragment(string theString)
        {
            value = theString;
        }
    }

    public class Integer : Literal
    {
        public string value;
        public Integer(string theString)
        {
            value = theString;
        }
    }

    public class FloatingPoint : Literal
    {
        public string value;
        public FloatingPoint(string theString)
        {
            value = theString;
        }
    }

    public class Hex : Literal
    {
        public string value;
        public Hex(string theString)
        {
            value = theString;
        }
    }

    public class LineFeed : Token
    {
    }

    public class Colon : Token
    {
    }

    public class Semicolon : Token
    {
    }

    public class Position
    {
        public ulong line;
        public ulong column;
    }
}
